libraries/chain/controller.cpp

```diff
@@ -4107,6 +4110,12 @@ struct controller_impl {
       }
    }
 
+   std::optional<finality_data_t> head_finality_data() const {
+      return apply<std::optional<finality_data_t>>(chain_head,
+         overloaded{ [](const block_state_legacy_ptr& head) { return std::nullopt; },
+                     [](const block_state_ptr&        head) { return head->get_finality_data(); }});
+   }
```

Comment:
```suggestion
      return apply_s<std::optional<finality_data_t>>(chain_head, [](const block_state_ptr& head) { return head->get_finality_data(); });
   }
```

Reply:
`apply_s` will return an empty `finality_data_t` in Legacy. `SHiP` cannot tell if it has a value or not and will proceed to  `"finality_data": null`. In Legacy, we don't want `SHiP` to generate the `finality_data` log at all so it won't break existing clients.

Reply:
> `apply_s` will return an empty `finality_data_t` in Legacy.

I'm pretty sure it it will return an empty optional.

Reply:
I initially used `apply_s` and tried again, the check https://github.com/AntelopeIO/leap/blob/417615de4cd035e962d8cbdf5719d517bbab8a6c/plugins/state_history_plugin/state_history_plugin.cpp#L283 always succeeded in Legacy tests.

Reply:
Which test would fail with this change?

Reply:
Like the legacy `ship_streamer_test` by `ctest -R ship_streamer_test`.

Reply:
The test itself does not fail, but an extra field `"finality_data": null` is added to SHiP log for Legacy, which is incorrect.

I will add a test to make sure no extra fields added for Legacy later.

Reply:
Looks like the extra field `"finality_data": null ` is added even with the current code in the branch.

Reply:
The extra field was caused by ABI binary to JSON converter, even though the SHiP binary log does not include `finality_data`. https://github.com/AntelopeIO/leap/issues/2340 is created to track the issue.
Thanks.

Reply:
Thanks for creating the new issue, Lin!